### PR TITLE
Type ``_infer`` of Attribute, AssignAttr, Global and Subscript

### PR DESCRIPTION
**Pull request by DanielNoord**
## Steps

- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Related Issue

### CODE REVIEW

**Review by cdce8p — APPROVED** [review]

**Review by Pierre-Sassoulas — APPROVED** [review]
LGTM, nice cleanup.

File: astroid/inference.py [hunk 1]
```diff
@@ -346,10 +344,7 @@ def infer_attribute(
             yield owner
             continue
 
-        if not context:
-            context = InferenceContext()
-        else:
-            context = copy_context(context)
+        context = copy_context(context)
```
Comment (cdce8p) [inline astroid/inference.py:348]:
```suggestion
```
Empty line

File: astroid/inference.py [hunk 2]
```diff
@@ -358,20 +358,20 @@ def infer_attribute(self, context=None):
...
 
 
 @decorators.raise_if_nothing_inferred
 @decorators.path_wrapper
-def infer_global(self, context=None):
-    if context.lookupname is None:
+def infer_global(
+    self: nodes.Global, context: InferenceContext | None = None, **kwargs: Any
+) -> Generator[InferenceResult, None, None]:
+    if not context or context.lookupname is None:
```
Comment (cdce8p) [inline astroid/inference.py:374]:
```suggestion
    if context is None or context.lookupname is None:
```

**Review by cdce8p — APPROVED** [review]

**Comment by cdce8p** [comment]
Somehow, this changes causes a pylint failure with Home Assistant.
https://github.com/cdce8p/ha-core/runs/7053862874?check_suite_focus=true#step:8:93

I'm currently running a test to see if a revert fixes the issue.-
https://github.com/cdce8p/ha-core/runs/7053960620?check_suite_focus=true

**Comment by DanielNoord** [comment]
Could it be that we actually depend on the crash of the missing `context`? And that we shouldn't create a new context there?

**Comment by cdce8p** [comment]
> Could it be that we actually depend on the crash of the missing `context`? And that we shouldn't create a new context there?

Could be. Will need to investigate that. For now, I would suggest to revert it completely. That does fix it https://github.com/cdce8p/ha-core/runs/7053960620?check_suite_focus=true

**Comment by cdce8p** [comment]
**Update**
Not sure why, but I narrowed it down to the decorator changes: https://github.com/PyCQA/astroid/commit/eb1b58e59a90dcd20f8180e8436109545ddbd8c2
The context modifications seem to be fine.

**Comment by DanielNoord** [comment]
> **Update**
> Not sure why, but I narrowed it down to the decorator changes: [eb1b58e](https://github.com/PyCQA/astroid/commit/eb1b58e59a90dcd20f8180e8436109545ddbd8c2)
> The context modifications seem to be fine.

Should we re-commit without the decorator changes?

**Comment by cdce8p** [comment]
> > **Update**
> > Not sure why, but I narrowed it down to the decorator changes: [eb1b58e](https://github.com/PyCQA/astroid/commit/eb1b58e59a90dcd20f8180e8436109545ddbd8c2)
> > The context modifications seem to be fine.
> 
> Should we re-commit without the decorator changes?

👍🏻 I can test Home Assistant before we merge it, just to be sure.

**Comment by DanielNoord** [comment]
Do you have any idea as to why this would break? Shouldn't the decorators work as decorators as well?

**Comment by cdce8p** [comment]
> Do you have any idea as to why this would break? Shouldn't the decorators work as decorators as well?

Just found this commit: https://github.com/PyCQA/astroid/commit/0a8a75db30da060a24922e05048bc270230f5bad
I think that explains it. Missed it myself during the review.

**Pull request merged**